### PR TITLE
[3/n] Add skip-softmax to Triton flash attention kernel

### PR DESCRIPTION
**Pull request by kaix-nv**
### What does this PR do?

Type of change: ? 

New feature. Add skip-softmax tile skipping to the Triton flash attention kernel.

### Usage

```python
# Add a code snippet demonstrating how to use this
from modelopt.torch.kernels import attention

# Skip-softmax with threshold 0.1 (tiles contributing < 10% are skipped)
out = attention(q, k, v, b_start_loc, b_seq_len, max_len,
                skip_softmax_threshold=0.1)

# Via mtsa.sparsify() on HuggingFace models
import modelopt.torch.sparsity.attention_sparsity as mtsa
from transformers import AutoModelForCausalLM

model = AutoModelForCausalLM.from_pretrained("meta-llama/Llama-3.1-8B",
                                              torch_dtype=torch.bfloat16,
                                              device_map="cuda")

# Default config
mtsa.sparsify(model, mtsa.SKIP_SOFTMAX_TRITON_DEFAULT)
```

### Testing

Performance (TFLOPS at seq_len=16384, RTX 6000 Pro):
| SEQ_LEN | ModelOpt Triton | PyTorch SDPA | Flash Attention 2 | Skip-Softmax t=0.01 | Skip-Softmax t=0.1 |
|---:|---:|---:|---:|---:|---:|
| 16384.0 | 188.849922 | 211.718193 | 224.242843 | 172.901804 | 279.861684 |
| 32768.0 | 175.321787 | 212.815740 | 224.833553 | 146.150702 | 262.490463 |
| 65536.0 | 167.302839 | 214.932407 | 226.456141 | 145.082937 | 243.344791 |

</body></html>

### Before your PR is "*Ready for review*"

Make sure you read and follow [Contributor guidelines](https://github.com/NVIDIA/Model-Optimizer/blob/main/CONTRIBUTING.md) and your commits are signed (`git commit -s -S`).

Make sure you read and follow the [Security Best Practices](https://github.com/NVIDIA/Model-Optimizer/blob/main/SECURITY.md#security-coding-practices-for-contributors) (e.g. avoiding hardcoded `trust_remote_code=True`, `torch.load(..., weights_only=False)`, `pickle`, etc.).

- Is this change backward compatible?: ✅ / ❌ / N/A 
- If you copied code from any other sources or added a new PIP dependency, did you follow guidance in `CONTRIBUTING.md`: ✅ / ❌ / N/A 
- Did you write any new necessary tests?: ✅ / ❌ / N/A 
- Did you update [Changelog](https://github.com/NVIDIA/Model-Optimizer/blob/main/CHANGELOG.rst)?: ✅ / ❌ / N/A 

## Summary by CodeRabbit

* **New Features**
  * Added a Triton "skip-softmax" tile-skipping option for flash attention with a new attention keyword and configurable threshold (default 0.1).
  * Added a new sparse attention method and a default sparse configuration that enables the Triton skip-softmax method.

* **Tests**
  * Added GPU tests covering threshold behavior, numerical fidelity vs dense, shape preservation, decode-mode, and integration with sparsify.

* **Documentation**
  * Updated changelog for the new feature and removed two prior listed entries.

### CODE REVIEW

File: modelopt/torch/kernels/triton_fa.py [hunk 1]
```diff
@@ -467,6 +507,12 @@ def _attn_bwd_dkdv(
             scores = _apply_mask(scores, q_pos, kv_pos, seq_len_q, seq_len_kv, kv_start, IS_CAUSAL)
             p = tl.math.exp2(scores - lse[:, None])
 
+            # Re-apply skip-softmax: zero out rows that were skipped in forward
+            if APPLY_SKIP_SOFTMAX:
+                tile_row_max = tl.max(scores, 1)
+                can_skip = tile_row_max < (lse + SKIP_THRESHOLD_LOG2)
```
Comment (Edwardf0t1) [inline modelopt/torch/kernels/triton_fa.py:732]: So forward and backward is different — forward skips based on row_max, backward on lse.

The lse >= row_max always holds, so the backward threshold is strictly looser — it will skip fewer tiles than the forward. This means the backward may compute gradients for tiles that were skipped in the forward.

Is this intentional?

File: modelopt/torch/kernels/triton_fa.py [hunk 2]
```diff
@@ -162,26 +164,54 @@ def _attn_fwd(
...
+            can_skip = tile_row_max < (row_max + SKIP_THRESHOLD_LOG2)
+            all_skip = tl.min(can_skip.to(tl.int32)) == 1
+
+            if not all_skip:
+                # Online softmax update (reuses tile_row_max — no second tl.max)
+                # For skipped rows: tile_row_max < row_max, so m_new = row_max (no change)
+                m_new = tl.maximum(row_max, tile_row_max)
+                p = tl.math.exp2(scores - m_new[:, None])
+                # Zero out skipped rows (instead of masking scores and recomputing max)
+                p = tl.where(can_skip[:, None], 0.0, p)
```
Comment (jingyu-ml) [inline modelopt/torch/kernels/triton_fa.py]: here's confusing, if not all_skip, we don't do the skip for this tile, then why we add the p = tl.where(can_skip[:, None], 0.0, p) to here since we skip nothing for this tile, and we are doing the tile level skipping

Comment (jingyu-ml) [reply modelopt/torch/kernels/triton_fa.py]: I guess all_skip means that, for this specific tile, it’s true if we skip it and false otherwise. If that’s the case, the variable name is confusing as well._

File: modelopt/torch/kernels/triton_fa.py [hunk 3]
```diff
@@ -162,26 +164,54 @@ def _attn_fwd(
...
+
+            # Load V and accumulate
+            v_offs = (kv_offset + kv_start + kv_pos[:, None]) * stride_vbs + dim_pos[None, :]
+            v = tl.load(
+                v_base + v_offs,
+                mask=((kv_start + kv_pos[:, None]) < seq_len_kv) & d_mask[None, :],
+                other=0.0,
+            )
+            acc = tl.dot(p.to(v.dtype), v, acc)
+            row_max = m_new
```
Comment (Edwardf0t1) [inline modelopt/torch/kernels/triton_fa.py:385]: It seems one path for APPLY_SKIP_SOFTMAX=True and one for False. The standard path and the "not all_skip" branch inside the skip path are nearly identical. Since APPLY_SKIP_SOFTMAX is a tl.constexpr, Triton will compile-time eliminate the dead branch, but the source code duplication makes maintenance harder. 

Consider structuring as:     
```  
# Skip check (compiled out when APPLY_SKIP_SOFTMAX=False)                                                                                                                                                                                                               
  do_process = True                                                                                                                                                                                                                                                       
  if APPLY_SKIP_SOFTMAX:                                                                                                                                                                                                                                                  
      tile_row_max = tl.max(scores, 1)                                                                                                                                                                                                                                    
      can_skip = tile_row_max < (row_max + SKIP_THRESHOLD_LOG2)                                                                                                                                                                                                           
      all_skip = tl.min(can_skip.to(tl.int32)) == 1                                                                                                                                                                                                                       
      if all_skip:                                                                                                                                                                                                                                                        
          do_process = False                                                                                                                                                                                                                                              
                                                                                                                                                                                                                                                                          
  if do_process:                                                                                                                                                                                                                                                          
      # Single copy of softmax update + V accumulation                                                                                                                                                                                                                    
      ...                                                                                                                                                                                                                                                                 
      if APPLY_SKIP_SOFTMAX:                                                                                                                                                                                                                                              
          p = tl.where(can_skip[:, None], 0.0, p)                                                                                                                                                                                                                         
      ...   
```

File: modelopt/torch/kernels/triton_fa.py (hunk 2 above)
Comment (jingyu-ml) [reply modelopt/torch/kernels/triton_fa.py]: seems to me, this line should be deleted

**Comment by kaix-nv** [comment]
@coderabbitai full review

**Comment by kaix-nv** [comment]
@coderabbitai full review

File: modelopt/torch/kernels/triton_fa.py (hunk 1 above)
Comment (kaix-nv) [reply modelopt/torch/kernels/triton_fa.py:732]: Good catch. This is intentional. It's not trivial to exactly reproduce the forward skip decisions because:
- Forward uses the running max at each tile step (not saved).
- Forward and backward may use different BLOCK_N (forward autotunes, backward uses fixed 64).

File: modelopt/torch/kernels/triton_fa.py (hunk 3 above)
Comment (kaix-nv) [reply modelopt/torch/kernels/triton_fa.py:385]: Good observation. The duplication is intentional. I considered this refactoring but Triton doesn't support reassigning a variable from a  `tl.constexpr` branch and using it as a runtime condition. The reassignment is silently ignored. If we remove `tl.constexpr`  to work around this, the runtime if/else adds branch divergence overhead on every attention tile iteration.

**Review by rohansjoshi — APPROVED** [review]
